# IST 718: Big Data Analytics

- Professor: Daniel Acuna <deacuna@syr.edu>
- TAs: Tong Zeng <tozeng@syr.edu>, Priya Matnani <psmatnan@syr.edu>

## General instructions:

- You are welcome to discuss the problems with your classmates but __you are not allowed to copy any part of your answers either from your classmates or from the internet__
- You can put the homework files anywhere you want in your http://notebook.acuna.io workspace but _do not change_ the file names. The TAs and the professor use these names to grade your homework.
- Remove or comment out code that contains `raise NotImplementedError`. This is mainly to make the `assert` statement fail if nothing is submitted.
- The tests shown in some cells (i.e., `assert` and `np.testing.` statements) are used to grade your answers. **However, the professor and TAs will use __additional__ test for your answer. Think about cases where your code should run even if it passess all the tests you see.**
- Before downloading and submitting your work through Blackboard, remember to save and press `Validate` (or go to 
`Kernel`$\rightarrow$`Restart and Run All`). 
- Good luck!

In [1]:
# load these packages
import pyspark
from pyspark.ml import feature, classification
from pyspark.ml import Pipeline
from pyspark.sql import functions as fn
import numpy as np
from pyspark.sql import SparkSession
from pyspark.ml import feature, regression, evaluation, Pipeline
from pyspark.sql import functions as fn, Row
import matplotlib.pyplot as plt
spark = SparkSession.builder.getOrCreate()
sc = spark.sparkContext
import pandas as pd

# Part 1: Random Forest and gradient boosted trees

In these questions, we will examine the famous [Auto dataset](https://vincentarelbundock.github.io/Rdatasets/doc/ISLR/Auto.html). With this dataset, the goal is to predict the miles per gallon (`mpg`) performance based on characteristics of the car such as number of cylinders (`cylinders`), displacement between wheels (`displacement`), horsepower of the engine (`horsepower`), weight of the car (`weight`), top acceleration (`acceleration`), year of the model (`year`), and origin (`origin`).

In [2]:
# read-only
mpg_df = spark.read.csv('/datasets/ISLR/Auto.csv', header=True, inferSchema=True).\
    drop('_c0').\
    withColumn('horsepower2', fn.col('horsepower').cast('int')).\
    drop('horsepower').\
    withColumnRenamed('horsepower2', 'horsepower').\
    dropna()
training_df, validation_df, testing_df = mpg_df.randomSplit([0.6, 0.3, 0.1], seed=0)
mpg_df.printSchema()

root
 |-- mpg: double (nullable = true)
 |-- cylinders: integer (nullable = true)
 |-- displacement: double (nullable = true)
 |-- weight: integer (nullable = true)
 |-- acceleration: double (nullable = true)
 |-- year: integer (nullable = true)
 |-- origin: integer (nullable = true)
 |-- name: string (nullable = true)
 |-- horsepower: integer (nullable = true)



# Question 1: (10 pts)

Create three pipelines that contain three different random forests that take in all features from `mpg_df` (`cylinders`, `displacement`, `horsepower`, `weight`, `acceleration`, `year`, and `origin`) to predict (`mpg`). **Set the `seed` parameter of the random forest to 0.** Fit these pipelines to the training data (`training_df`):

- `pipe_rf1`: Random forest with `maxDepth=1` and `numTrees=60`
- `pipe_rf2`: Random forest with `maxDepth=3` and `numTrees=40`
- `pipe_rf3`: Random forest with `maxDepth=6`, `numTrees=20`

In [3]:
# create the fitted pipelines `pipe_rf1`, `pipe_rf2`, and `pipe_rf3` here
pipe_rf1 = Pipeline(stages=[feature.VectorAssembler(inputCols=['cylinders','displacement','horsepower','weight',
                                                               'acceleration','year','origin'], outputCol='feature'),
                            regression.RandomForestRegressor(numTrees=60,seed=0,featureSubsetStrategy='auto',maxDepth=1,labelCol='mpg',featuresCol='feature')]).fit(training_df)

pipe_rf2 = Pipeline(stages=[feature.VectorAssembler(inputCols=['cylinders','displacement','horsepower','weight',
                                                               'acceleration','year','origin'], outputCol='feature'),
                            regression.RandomForestRegressor(numTrees=40,seed=0,featureSubsetStrategy='auto',maxDepth=3,labelCol='mpg',featuresCol='feature')]).fit(training_df)

pipe_rf3 = Pipeline(stages=[feature.VectorAssembler(inputCols=['cylinders','displacement','horsepower','weight',
                                                               'acceleration','year','origin'], outputCol='feature'),
                            regression.RandomForestRegressor(numTrees=20,seed=0,featureSubsetStrategy='auto',maxDepth=6,labelCol='mpg',featuresCol='feature')]).fit(training_df)

In [4]:
training_df.show(1)

+----+---------+------------+------+------------+----+------+---------+----------+
| mpg|cylinders|displacement|weight|acceleration|year|origin|     name|horsepower|
+----+---------+------------+------+------------+----+------+---------+----------+
|10.0|        8|       307.0|  4376|        15.0|  70|     1|chevy c20|       200|
+----+---------+------------+------+------------+----+------+---------+----------+
only showing top 1 row



In [5]:
# tests for 10 pts
np.testing.assert_equal(type(pipe_rf1.stages[0]), feature.VectorAssembler)
np.testing.assert_equal(type(pipe_rf2.stages[0]), feature.VectorAssembler)
np.testing.assert_equal(type(pipe_rf3.stages[0]), feature.VectorAssembler)
np.testing.assert_equal(type(pipe_rf1.stages[1]), regression.RandomForestRegressionModel)
np.testing.assert_equal(type(pipe_rf2.stages[1]), regression.RandomForestRegressionModel)
np.testing.assert_equal(type(pipe_rf3.stages[1]), regression.RandomForestRegressionModel)
np.testing.assert_equal(type(pipe_rf1.transform(training_df)), pyspark.sql.dataframe.DataFrame)
np.testing.assert_equal(type(pipe_rf2.transform(training_df)), pyspark.sql.dataframe.DataFrame)
np.testing.assert_equal(type(pipe_rf3.transform(training_df)), pyspark.sql.dataframe.DataFrame)

# Question 2 (10 pts)

Use the following evaluator to compute the $R^2$ of the models on validation data. Assign the $R^2$ of the three models to `R2_1`, `R2_2`, and `R2_3`, respectively, and the performance. Assign the best pipeline based on validation performance to a variable `best_model`

In [6]:
evaluator = evaluation.RegressionEvaluator(labelCol='mpg', metricName='r2')
# use it as follows:
#   evaluator.evaluate(fitted_pipeline.transform(df)) -> R2

In [7]:
# YOUR CODE HERE
#pipe_rf1.transform(validation_df)
R2_1 = evaluator.evaluate(pipe_rf1.transform(validation_df))
R2_2 = evaluator.evaluate(pipe_rf2.transform(validation_df))
R2_3 = evaluator.evaluate(pipe_rf3.transform(validation_df))

print("R2_1:",R2_1,"R2_2:",R2_2,"R2_3:",R2_3)

best_model = pipe_rf3

R2_1: 0.6336763409868249 R2_2: 0.8404115955766471 R2_3: 0.8799114180154942


In [8]:
# tests for 10 pts
np.testing.assert_equal(type(best_model.stages[0]), feature.VectorAssembler)
np.testing.assert_equal(type(best_model.stages[1]), regression.RandomForestRegressionModel)
np.testing.assert_equal(type(best_model.transform(validation_df)), pyspark.sql.dataframe.DataFrame)
np.testing.assert_array_less(R2_1, 1.)
np.testing.assert_array_less(0.5, R2_1)
np.testing.assert_array_less(R2_2, 1.)
np.testing.assert_array_less(0.5, R2_2)
np.testing.assert_array_less(R2_3, 1.)
np.testing.assert_array_less(0.5, R2_3)

# Question 3: 5 pts

Compute the $R^2$ of the model on testing data, print it, and assign it to variable `R2_best`

In [9]:
# create AUC_best below
# YOUR CODE HERE
R2_best = evaluator.evaluate(best_model.transform(testing_df))
R2_best

0.9664636858674139

In [10]:
# tests for 5 pts
np.testing.assert_array_less(R2_best, 1.)
np.testing.assert_array_less(0.5, R2_best)

# Question 4: 5 pts

Using the parameters of the best model, create a new pipeline called `final_model` and fit it to the entire data (`mpg_df`)

In [11]:
# create the fitted pipeline `final_model` here
# YOUR CODE HERE
final_model =  Pipeline(stages=[feature.VectorAssembler(inputCols=['cylinders','displacement','horsepower','weight',
                                                               'acceleration','year','origin'], outputCol='feature'),
                            regression.RandomForestRegressor(numTrees=20,seed=0,featureSubsetStrategy='auto',maxDepth=6,labelCol='mpg',featuresCol='feature')]).fit(mpg_df)

In [12]:
# tests for 10 pts
np.testing.assert_equal(type(final_model.stages[0]), feature.VectorAssembler)
np.testing.assert_equal(type(final_model.stages[1]), regression.RandomForestRegressionModel)
np.testing.assert_equal(type(final_model.transform(mpg_df)), pyspark.sql.dataframe.DataFrame)

# Question 5: 10 pts

Create a pandas dataframe `feature_importance` with the columns `feature` and `importance` which contains the names of the features (`cylinder`, `displacement`, etc.) and their feature importances as determined by the random forest of the final model. Sort the dataframe by `importance` in descending order.

In [13]:
# create feature_importance below
# YOUR CODE HERE
feature = ['cylinders','displacement','horsepower','weight','acceleration','year','origin']

feature_importance = pd.DataFrame(list(zip(feature,final_model.stages[-1].featureImportances.toArray())),
            columns = ['feature', 'importance']).sort_values('importance', ascending = False)

In [14]:
# display it here
feature_importance

feature  importance
0     cylinders    0.258947
1  displacement    0.209820
3        weight    0.186403
2    horsepower    0.156602
5          year    0.156356
4  acceleration    0.023101
6        origin    0.008771

In [15]:
# tests for 10 pts
assert type(feature_importance) == pd.core.frame.DataFrame
np.testing.assert_array_equal(list(feature_importance.columns), ['feature', 'importance'])
np.testing.assert_array_equal(list(feature_importance.columns), ['feature', 'importance'])

**(5 pts)** Comment below on the importance that random forest has given to each feature. Are they reasonable? Do they tell you anything valuable about the mpg dataset? Answer in the cell below

### Comment
Miles per gallon reflects the efficiency of the car. If we want to make a car to be efficient, cylinders play the most crucial role and that is what the feature importance is showing. The displacement of engine is also almost as important as the cylinders for efficieny of the car. After the above two features, weight of car being important also makes sense. The more the weight of the car, the fuel consumption should be also impacted. In contrast to the above, origin of the car i.e American, European, Japanese will not impact the fuel consumption that much. The term acceleration of a car is more related to performance than to efficiency, therefore its positioning in the features makes sense.
The remaining two words horsepower and year might impact the efficiency of the car but not as cucial as displacement, and cylinders.

# Question 6:  5 pts.
`
Pick any of the trees from the final model and assign its `toDebugString` property to a variable `example_tree`. Print this variable and add comments to the cell describing how you think this particular tree is fitting the data

In [16]:
# create a variable example_tree with the toDebugString property of a tree from final_model.
# print this string and comment in this same cell about the branches that this tree fit
# YOUR CODE HERE
rfModel = final_model.stages[-1]
len(rfModel.trees)
example_tree = rfModel.trees[2].toDebugString
print(example_tree)

DecisionTreeRegressionModel (uid=dtr_baa3165fedb7) of depth 6 with 117 nodes
  If (feature 0 <= 4.5)
   If (feature 5 <= 78.5)
    If (feature 2 <= 71.5)
     If (feature 5 <= 76.5)
      If (feature 6 <= 2.5)
       If (feature 4 <= 19.55)
        Predict: 29.0
       Else (feature 4 > 19.55)
        Predict: 26.818181818181817
      Else (feature 6 > 2.5)
       If (feature 2 <= 53.5)
        Predict: 32.6
       Else (feature 2 > 53.5)
        Predict: 31.666666666666668
     Else (feature 5 > 76.5)
      If (feature 3 <= 2115.0)
       If (feature 4 <= 19.55)
        Predict: 34.11538461538461
       Else (feature 4 > 19.55)
        Predict: 43.10000000000002
      Else (feature 3 > 2115.0)
       If (feature 6 <= 1.5)
        Predict: 30.0
       Else (feature 6 > 1.5)
        Predict: 29.5
    Else (feature 2 > 71.5)
     If (feature 1 <= 107.5)
      If (feature 0 <= 3.5)
       If (feature 1 <= 79.5)
        Predict: 18.0
       Else (feature 1 > 79.5)
        Predict: 21.5
   

In [17]:
final_model.transform(testing_df).take(1)

[Row(mpg=11.0, cylinders=8, displacement=400.0, weight=4997, acceleration=14.0, year=73, origin=1, name='chevrolet impala', horsepower=150, feature=DenseVector([8.0, 400.0, 150.0, 4997.0, 14.0, 73.0, 1.0]), prediction=13.351007719380604)]

### Comment
Comparing one testing prediction with Decision Tree,

(f0)cylinders=8                      here f0 is greater than 4.5
(f2)horsepower=150,                  here f2 is greater than 138.5
(f3)weight=4997,                     here f3 is greater than 3462
(f5)year=73,                         here f5 is less than 76.5
(f4)acceleration=14.0,               here f4 is greater than 13.75
Therefore predicted value should be 13.55
By using the testing data the predicted value is coming 13.35
Therefore, this tree is fitting the data considerably.



In [18]:
# tests for 5 points
assert type(example_tree) == str
assert 'DecisionTreeRegressionModel' in example_tree
assert 'feature 0' in example_tree
assert 'If' in example_tree
assert 'Else' in example_tree
assert 'Predict' in example_tree

# **Question 7 (5 pts)**

Gradient boosted trees are becoming increasingly popular for competitions. There is a high-performance implementation, [xgboost](https://en.wikipedia.org/wiki/XGBoost), that is particularly popular. Compare gradient boosted regression to the best model found with random forest in Question 3. Use the validation set. For GBR, use all the default parameters except make `seed=0`. Assign the pipeline and the $R^2$ of the model to `gbr_pipe` and `R2_gbr`, respectively. Does it have an amazing or dissapointing $R^2$? Comment.

In [19]:
# YOUR CODE HERE
from pyspark.ml import feature, classification
gbr_pipe = Pipeline(stages=[feature.VectorAssembler(inputCols=['cylinders','displacement','horsepower','weight','acceleration','year','origin'], 
                                                    outputCol='feature'), regression.GBTRegressor(featuresCol="feature",labelCol='mpg', seed=0)]).fit(training_df)
R2_gbr = evaluator.evaluate(gbr_pipe.transform(validation_df))
R2_gbr

0.852350432034137

In [20]:
# test your models here
print("Performance of best RF: ", evaluator.evaluate(best_model.transform(validation_df)))
print("Performance of GBR: ", R2_gbr)

Performance of best RF:  0.8799114180154942
Performance of GBR:  0.852350432034137


In [21]:
# tests for 5 pts
np.testing.assert_equal(type(gbr_pipe.stages[0]), feature.VectorAssembler)
np.testing.assert_equal(type(gbr_pipe.stages[1]), regression.GBTRegressionModel)
np.testing.assert_equal(type(gbr_pipe.transform(validation_df)), pyspark.sql.dataframe.DataFrame)
np.testing.assert_array_less(R2_gbr, 1.)
np.testing.assert_array_less(0.5, R2_gbr)

### Comment
The r sqaure of Random Forest is better as compared to gradient boosting regressor marginally. The R square value is good for both Random forest as well Gradient boosting regressor, however it is not amazing.